### PR TITLE
fix certificate mismatch validation logic

### LINKED ISSUE #76: Certificate matching doesn't seem to be implemented correctly
Reference: https://datatracker.ietf.org/doc/html/rfc6125#section-6.4.3

Initial tests are implemented at https://github.com/projectdiscovery/tlsx/commit/c5065af7e3dc98a2c662f15d7f0d59a76444a781 with few commented-out tests that are not covered with the current implementation and need to be fixed.

```
		/* TODO enable these tests once the functionality is corrected #76
		{args{host: "baz1.example.net", names: []string{"baz*.example.net"}}, false},
		{args{host: "foobaz.example.net", names: []string{"*baz.example.net"}}, false},
		{args{host: "buzz.example.net", names: []string{"b*z.example.net"}}, false},
		*/

		// negative scenarios
		/* TODO enable these tests once the functionality is corrected #76
		{args{host: "subdomain.target.com", names: []string{"other-target.com", "subdomain.target.*"}}, true},
		{args{host: "subdomain.target.com", names: []string{"subdomain.*.com", "other-target.com"}}, true},
		*/
```

- [x] Fix current implementation to cover all the cases
- [x] Uncomment tests after fixing and ensure all tests are passing.

### PR DESCRIPTION
**Pull request by parrasajad**
fixes #76 , tests cases are passing now

### CODE REVIEW

**Review by Ice3man543 — APPROVED** [review]

File: pkg/tlsx/clients/clients_test.go [hunk 1]
```diff
@@ -51,3 +47,31 @@ func TestIsMisMatchedCert(t *testing.T) {
...
+		{name: "wildcard00", args: args{nameToken: "b*z", hostToken: "buzz"}, want: true},
+		{name: "wildcard01", args: args{nameToken: "*buzz", hostToken: "foobuzz"}, want: true},
+		{name: "wildcard02", args: args{nameToken: "foo*", hostToken: "foobuzz"}, want: true},
+		{name: "wildcard03", args: args{nameToken: "foo*", hostToken: "buzz"}, want: false},
+		{name: "wildcard04", args: args{nameToken: "*buzz", hostToken: "foo"}, want: false},
+		{name: "wildcard05", args: args{nameToken: "*", hostToken: "foo"}, want: true},
+		{name: "wildcard06", args: args{nameToken: "subdomain", hostToken: "subdomain"}, want: true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (parrasajad) [reply pkg/tlsx/clients/clients_test.go:71]: i found this useful/faster when trying to find the failing test :)

File: pkg/tlsx/clients/clients_test.go [hunk 2]
```diff
@@ -51,3 +47,31 @@ func TestIsMisMatchedCert(t *testing.T) {
...
+			if got := matchWildCardToken(tt.args.nameToken, tt.args.hostToken); got != tt.want {
+				t.Errorf("matchWildCardToken() = %v, want %v", got, tt.want)
```
[the remark this answers is not in the document]
Comment (parrasajad) [reply pkg/tlsx/clients/clients_test.go:73]: tried to follow the existing test present in the file.

File: pkg/tlsx/clients/clients_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (forgedhallpass) [reply pkg/tlsx/clients/clients_test.go:71]: Have you tried my suggestion? :) The first part literally does the same thing that you have right now, without hard-coding the name, and the second suggestion explicitly shows which test has failed, so you don't even need to look it up.

File: pkg/tlsx/clients/clients_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (forgedhallpass) [reply pkg/tlsx/clients/clients_test.go:73]: That doesn't mean that it cannot be done better. It's exactly why there are assertion functions.

**Review by forgedhallpass — CHANGES_REQUESTED** [review]
See my comments

**Pull request merged**